File: go/youngestCommonAncestor/youngestCommonAncestor.go
package main


import "fmt"


type AncestralTree struct {
	Value int
	Ancestor *AncestralTree
}


// Time -> 0(log(N)) Space -> 0(log(N))
func getYoungestCommonAncestor(top, node1, node2 *AncestralTree) *AncestralTree {
	ancestors := node1
	ancestor_list := make(map[*AncestralTree]bool)
	for ancestors != nil {
		ancestor_list[ancestors] = false
		ancestors = ancestors.Ancestor
	}

	for node2 != nil {
		if _, exists := ancestor_list[node2]; exists {
			return node2
		}
		node2 = node2.Ancestor
	}

	return top
}

// Time -> 0(log(N)) Space -> 0(1)
func getYoungestCommonAncestor1(top, node1, node2 * AncestralTree) *AncestralTree {
	var node1Count, node2Count int
	node1Tree := node1
	node2Tree := node2
	for node1Tree != nil {
		node1Count++
		node1Tree = node1Tree.Ancestor
	}
	for node2Tree != nil {
		node2Count++
		node2Tree = node2Tree.Ancestor
	}

	if node1Count > node2Count {
		return levelTree(node1, node2, node1Count-node2Count)
	}else {
		return levelTree(node2, node1, node2Count-node1Count)
	}

}


func levelTree(node, highnode *AncestralTree, diff int) *AncestralTree {
	levelNode := node
	for diff > 0 {
		levelNode = levelNode.Ancestor
		diff--
	}
	
	for levelNode != highnode{
		levelNode = levelNode.Ancestor
		highnode = highnode.Ancestor
	}
	return levelNode
}

func main() {
	top := AncestralTree{Value: 1}
	a := AncestralTree{Value: 2, Ancestor: &top}
	b := AncestralTree{Value: 3, Ancestor: &top}
	node1 := AncestralTree{Value: 4, Ancestor: &a}
	node2 := AncestralTree{Value: 7, Ancestor: &b}
	fmt.Println(getYoungestCommonAncestor(&top, &node1, &node2).Value)
}

File: python/youngestCommonAncestor.py
class AncestryTree():
    def __init__(self, value, ancestor=None):
        self.value = value
        self.ancestor = ancestor
    

# Time -> 0(log(N)) Space -> 0(log(N))
def getYoungestCommonAncestor(top: AncestryTree, node1: AncestryTree, node2: AncestryTree) -> AncestryTree:
    ancestor_list = {}
    while node1 is not None:
        ancestor_list[node1] = True
        node1 = node1.ancestor
    while node2 is not None:
        if node2 in ancestor_list:
            return node2
        node2 = node2.ancestor
    return top


# Time -> 0(log(N)) Space -> 0(1)
def getYoungestCommonAncestor1(top: AncestryTree, node1: AncestryTree, node2: AncestryTree) -> AncestryTree:
    node1_count = node1
    node2_count = node2
    node1_c, node2_c = 0, 0
    while node1_count != None:
        node1_c += 1
        node1_count = node1_count.ancestor
    while node2_count != None:
        node2_c += 1
        node2_count = node2_count.ancestor
    
    if node1_c > node2_c:
        return level(node1, node2, node1_c-node2_c)
    else:
        return level(node2, node1, node2_c-node1_c)
    
    
def level(high: AncestryTree, level: AncestryTree, diff: int) -> AncestryTree:
    while diff > 0:
        diff -= 1
        high = high.ancestor
    
    while high != level:
        high = high.ancestor
        level = level.ancestor
    return level

top = AncestryTree(value=1)
a = AncestryTree(value=2, ancestor=top)
b = AncestryTree(value=3, ancestor=top)
node1 = AncestryTree(value=4, ancestor=a)
node2 = AncestryTree(value=7, ancestor=top)
print(getYoungestCommonAncestor1(top, node1, node2).value)